Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs

```diff
@@ -22,35 +22,51 @@ module Domain.Action.UI.Ride.CancelRide
 where
 
 import qualified Domain.Action.UI.Ride.CancelRide.Internal as CInternal
+import qualified Domain.Types.Booking as SRB
 import qualified Domain.Types.BookingCancellationReason as DBCR
 import Domain.Types.CancellationReason (CancellationReasonCode (..))
+import qualified Domain.Types.DriverLocation as DDriverLocation
+import Domain.Types.Merchant
 import qualified Domain.Types.Merchant as DM
 import qualified Domain.Types.Person as DP
 import qualified Domain.Types.Ride as DRide
 import Environment
+import Kernel.External.Maps
 import Kernel.Prelude
+import qualified Kernel.Storage.Esqueleto as Esq
 import qualified Kernel.Types.APISuccess as APISuccess
 import Kernel.Types.Common
 import Kernel.Types.Id
 import Kernel.Utils.Common
+import qualified SharedLogic.DriverLocation as QDrLoc
+import Storage.CachedQueries.CacheConfig
+import qualified Storage.Queries.Booking as QRB
 import qualified Storage.Queries.Person as QPerson
 import qualified Storage.Queries.Ride as QRide
 import Tools.Error
+import qualified Tools.Maps as Maps
+import qualified Tools.Metrics as Metrics
 
 type MonadHandler m = (MonadThrow m, Log m, MonadGuid m)
 
 data ServiceHandle m = ServiceHandle
   { findRideById :: Id DRide.Ride -> m (Maybe DRide.Ride),
     findById :: Id DP.Person -> m (Maybe DP.Person),
-    cancelRide :: Id DRide.Ride -> DBCR.BookingCancellationReason -> m ()
+    findDriverLocationId :: Id DP.Person -> m (Maybe DDriverLocation.DriverLocation),
+    cancelRide :: Id DRide.Ride -> DBCR.BookingCancellationReason -> m (),
+    findBookingByIdInReplica :: Id SRB.Booking -> m (Maybe SRB.Booking),
+    pickUpDistance :: Id DM.Merchant -> LatLong -> LatLong -> m Meters
   }
 
 cancelRideHandle :: ServiceHandle Flow
 cancelRideHandle =
   ServiceHandle
     { findRideById = QRide.findById,
       findById = QPerson.findById,
-      cancelRide = CInternal.cancelRideImpl
+      cancelRide = CInternal.cancelRideImpl,
+      findDriverLocationId = QDrLoc.findById,
+      findBookingByIdInReplica = Esq.runInReplica . QRB.findById,
+      pickUpDistance = getDistanceBetweenPoints
```

Comment:
Need not to create handler key `getDistanceBetweenPoints` function as it being called internally only.